src/systolic_array/commit.cpp

```diff
@@ -68,9 +68,17 @@ void Commit::evaluate() {
   // PE row is left idle. 2) Some of the PE columns are left idle due to a lack
   // of weights. In this case, we should do a writeback once all the "active"
   // columns have produced outputs.
-  for (int i = 0; i < ceil(float(inputs.size()) / elemsPerLine); i++) {
-    for (int j = 0; j < elemsPerLine && j < inputs.size(); j++) {
-      int index = i * elemsPerLine + j;
+
+  for (int remainingElems = inputs.size(); remainingElems > 0;
+       remainingElems -= elemsPerLine) {
+    // Check if we have collected all the pixels for a writeback.
+    int elemsToWrite = std::min(elemsPerLine, remainingElems);
+    int index = inputs.size() - remainingElems;
+    if (isLineComplete(index, elemsToWrite))
+      queueCommitRequest(index, elemsToWrite);
```

Comment:
Is moving this up here functionally equivalent? IsLineComplete reads outputBuffer, which is modified by the loop below.

Reply:
Yes it's functionally equivalent but I figured moving this up here is better in terms of modeling the hardware. Yes outputBuffer is modified by the loop below, but the hardware should see that and create a commit request in the _next_ cycle (if all elements in the line are collected).